File: python/chromahash/_decode.py
"""ChromaHash decoder. Per spec §11."""

from ._aspect import decode_output_size, derive_grid
from ._bitpack import read_bits
from ._color import linear_to_srgb8, oklab_to_linear_srgb, soft_gamut_clamp
from ._constants import (
    MAX_A_ALPHA_SCALE,
    MAX_A_SCALE,
    MAX_B_SCALE,
    MAX_CHROMA_A,
    MAX_CHROMA_B,
    MAX_L_SCALE,
)
from ._dct import dct_decode_pixel, triangular_scan_order
from ._math_utils import clamp01, round_half_away_from_zero
from ._mulaw import mu_law_dequantize


def decode(hash_bytes: bytes) -> tuple[int, int, bytes]:
    """Decode a ChromaHash into RGBA pixel data. Per spec §11.
    Returns (width, height, rgba_pixels).
    """
    # 1. Unpack header (48 bits)
    header = 0
    for i in range(6):
        header |= hash_bytes[i] << (i * 8)

    l_dc_q = header & 0x7F
    a_dc_q = (header >> 7) & 0x7F
    b_dc_q = (header >> 14) & 0x7F
    l_scl_q = (header >> 21) & 0x3F
    a_scl_q = (header >> 27) & 0x3F
    b_scl_q = (header >> 33) & 0x1F
    aspect = (header >> 38) & 0xFF
    has_alpha = ((header >> 46) & 1) == 1

    # 2. Decode DC values and scale factors
    l_dc = l_dc_q / 127.0
    a_dc = (a_dc_q - 64.0) / 63.0 * MAX_CHROMA_A
    b_dc = (b_dc_q - 64.0) / 63.0 * MAX_CHROMA_B
    l_scale = l_scl_q / 63.0 * MAX_L_SCALE
    a_scale = a_scl_q / 63.0 * MAX_A_SCALE
    b_scale = b_scl_q / 31.0 * MAX_B_SCALE

    # 3-4. Decode aspect ratio and compute output size
    w, h = decode_output_size(aspect)

    # 5. Dequantize AC coefficients
    bitpos = 48

    if has_alpha:
        alpha_dc_val = read_bits(hash_bytes, bitpos, 5) / 31.0
        bitpos += 5
        alpha_scale_val = read_bits(hash_bytes, bitpos, 4) / 15.0 * MAX_A_ALPHA_SCALE
        bitpos += 4
    else:
        alpha_dc_val = 1.0
        alpha_scale_val = 0.0

    if has_alpha:
        l_ac: list[float] = []
        for _ in range(7):
            q = read_bits(hash_bytes, bitpos, 6)
            bitpos += 6
            l_ac.append(mu_law_dequantize(q, 6) * l_scale)
        for _ in range(7, 20):
            q = read_bits(hash_bytes, bitpos, 5)
            bitpos += 5
            l_ac.append(mu_law_dequantize(q, 5) * l_scale)
    else:
        l_ac = []
        for _ in range(27):
            q = read_bits(hash_bytes, bitpos, 5)
            bitpos += 5
            l_ac.append(mu_law_dequantize(q, 5) * l_scale)

    a_ac: list[float] = []
    for _ in range(9):
        q = read_bits(hash_bytes, bitpos, 4)
        bitpos += 4
        a_ac.append(mu_law_dequantize(q, 4) * a_scale)

    b_ac: list[float] = []
    for _ in range(9):
        q = read_bits(hash_bytes, bitpos, 4)
        bitpos += 4
        b_ac.append(mu_law_dequantize(q, 4) * b_scale)

    if has_alpha:
        alpha_ac: list[float] = []
        for _ in range(5):
            q = read_bits(hash_bytes, bitpos, 4)
            bitpos += 4
            alpha_ac.append(mu_law_dequantize(q, 4) * alpha_scale_val)
    else:
        alpha_ac = []

    # Derive adaptive grid and compute usable scan orders (v0.2)
    l_dec_cap = 20 if has_alpha else 27
    l_nx, l_ny = derive_grid(aspect, 6 if has_alpha else 7)
    c_nx, c_ny = derive_grid(aspect, 4)

    l_scan_full = triangular_scan_order(l_nx, l_ny)
    l_usable = min(l_dec_cap, len(l_scan_full))
    l_scan = l_scan_full[:l_usable]
    l_ac_used = l_ac[:l_usable]

    chroma_scan_full = triangular_scan_order(c_nx, c_ny)
    c_usable = min(9, len(chroma_scan_full))
    chroma_scan = chroma_scan_full[:c_usable]
    a_ac_used = a_ac[:c_usable]
    b_ac_used = b_ac[:c_usable]

    alpha_scan: list = []
    alpha_ac_used: list[float] = []
    if has_alpha:
        a_nx, a_ny = derive_grid(aspect, 3)
        alpha_scan_full = triangular_scan_order(a_nx, a_ny)
        a_usable = min(5, len(alpha_scan_full))
        alpha_scan = alpha_scan_full[:a_usable]
        alpha_ac_used = alpha_ac[:a_usable]

    # 6. Render output image
    rgba = bytearray(w * h * 4)

    for y in range(h):
        for x in range(w):
            ok_l = dct_decode_pixel(l_dc, l_ac_used, l_scan, x, y, w, h)
            ok_a = dct_decode_pixel(a_dc, a_ac_used, chroma_scan, x, y, w, h)
            ok_b = dct_decode_pixel(b_dc, b_ac_used, chroma_scan, x, y, w, h)
            alpha = (
                dct_decode_pixel(alpha_dc_val, alpha_ac_used, alpha_scan, x, y, w, h)
                if has_alpha
                else 1.0
            )

            # Clamp L from DCT ringing, soft gamut clamp (v0.2)
            l_clamped = clamp01(ok_l)
            l_out, a_out, b_out = soft_gamut_clamp(l_clamped, ok_a, ok_b)
            rgb_lin = oklab_to_linear_srgb([l_out, a_out, b_out])
            idx = (y * w + x) * 4
            rgba[idx] = linear_to_srgb8(clamp01(rgb_lin[0]))
            rgba[idx + 1] = linear_to_srgb8(clamp01(rgb_lin[1]))
            rgba[idx + 2] = linear_to_srgb8(clamp01(rgb_lin[2]))
            rgba[idx + 3] = int(round_half_away_from_zero(255.0 * clamp01(alpha)))

    return (w, h, bytes(rgba))


def average_color(hash_bytes: bytes) -> tuple[int, int, int, int]:
    """Extract the average color without full decode. Per spec §11.2.
    Returns (r, g, b, a) as int values in [0, 255].
    """
    header = 0
    for i in range(6):
        header |= hash_bytes[i] << (i * 8)

    l_dc_q = header & 0x7F
    a_dc_q = (header >> 7) & 0x7F
    b_dc_q = (header >> 14) & 0x7F
    has_alpha = ((header >> 46) & 1) == 1

    l_dc = l_dc_q / 127.0
    a_dc = (a_dc_q - 64.0) / 63.0 * MAX_CHROMA_A
    b_dc = (b_dc_q - 64.0) / 63.0 * MAX_CHROMA_B

    # Apply soft gamut clamp to DC values (v0.2)
    l_clamped = clamp01(l_dc)
    l_out, a_out, b_out = soft_gamut_clamp(l_clamped, a_dc, b_dc)
    rgb_lin = oklab_to_linear_srgb([l_out, a_out, b_out])

    alpha = read_bits(hash_bytes, 48, 5) / 31.0 if has_alpha else 1.0

    return (
        linear_to_srgb8(clamp01(rgb_lin[0])),
        linear_to_srgb8(clamp01(rgb_lin[1])),
        linear_to_srgb8(clamp01(rgb_lin[2])),
        int(round_half_away_from_zero(255.0 * clamp01(alpha))),
    )

File: csharp/src/Chromahash/Decode.cs
namespace ChromaHash;

using static Constants;
using static MathUtils;

internal static class Decoder
{
    /// <summary>Decode a ChromaHash into RGBA pixel data. Per spec §11. Returns (width, height, rgba_pixels).</summary>
    public static (uint W, uint H, byte[] Rgba) Decode(byte[] hash)
    {
        // 1. Unpack header (48 bits)
        ulong header = 0;
        for (int i = 0; i < 6; i++)
            header |= ((ulong)hash[i]) << (i * 8);

        uint lDcQ = (uint)(header & 0x7F);
        uint aDcQ = (uint)((header >> 7) & 0x7F);
        uint bDcQ = (uint)((header >> 14) & 0x7F);
        uint lSclQ = (uint)((header >> 21) & 0x3F);
        uint aSclQ = (uint)((header >> 27) & 0x3F);
        uint bSclQ = (uint)((header >> 33) & 0x1F);
        byte aspect = (byte)((header >> 38) & 0xFF);
        bool hasAlpha = ((header >> 46) & 1) == 1;

        // 2. Decode DC values and scale factors
        double lDc = lDcQ / 127.0;
        double aDc = (aDcQ - 64.0) / 63.0 * MaxChromaA;
        double bDc = (bDcQ - 64.0) / 63.0 * MaxChromaB;
        double lScale = lSclQ / 63.0 * MaxLScale;
        double aScale = aSclQ / 63.0 * MaxAScale;
        double bScale = bSclQ / 31.0 * MaxBScale;

        // 3-4. Decode aspect ratio and compute output size
        (uint w, uint h) = Aspect.DecodeOutputSize(aspect);

        // 5. Dequantize AC coefficients
        int bitpos = 48;

        double alphaDcVal = 1.0;
        double alphaScaleVal = 0.0;

        if (hasAlpha)
        {
            alphaDcVal = BitPack.ReadBits(hash, bitpos, 5) / 31.0;
            bitpos += 5;
            alphaScaleVal = BitPack.ReadBits(hash, bitpos, 4) / 15.0 * MaxAlphaScale;
            bitpos += 4;
        }

        // Derive adaptive grid and compute usable scan orders (v0.2)
        int lDecCap = hasAlpha ? 20 : 27;
        (int lNx, int lNy) = Aspect.DeriveGrid(aspect, hasAlpha ? 6 : 7);
        (int cNx, int cNy) = Aspect.DeriveGrid(aspect, 4);

        var lScanFull = Dct.TriangularScanOrder(lNx, lNy);
        int lUsable = Math.Min(lDecCap, lScanFull.Count);

        List<double> lAc;
        if (hasAlpha)
        {
            lAc = new List<double>(20);
            for (int i = 0; i < 7; i++)
            {
                uint q = BitPack.ReadBits(hash, bitpos, 6);
                bitpos += 6;
                lAc.Add(MuLaw.MuLawDequantize(q, 6) * lScale);
            }
            for (int i = 7; i < 20; i++)
            {
                uint q = BitPack.ReadBits(hash, bitpos, 5);
                bitpos += 5;
                lAc.Add(MuLaw.MuLawDequantize(q, 5) * lScale);
            }
        }
        else
        {
            lAc = new List<double>(27);
            for (int i = 0; i < 27; i++)
            {
                uint q = BitPack.ReadBits(hash, bitpos, 5);
                bitpos += 5;
                lAc.Add(MuLaw.MuLawDequantize(q, 5) * lScale);
            }
        }

        var chromaScanFull = Dct.TriangularScanOrder(cNx, cNy);
        int cUsable = Math.Min(9, chromaScanFull.Count);

        var aAc = new List<double>(9);
        for (int i = 0; i < 9; i++)
        {
            uint q = BitPack.ReadBits(hash, bitpos, 4);
            bitpos += 4;
            aAc.Add(MuLaw.MuLawDequantize(q, 4) * aScale);
        }

        var bAc = new List<double>(9);
        for (int i = 0; i < 9; i++)
        {
            uint q = BitPack.ReadBits(hash, bitpos, 4);
            bitpos += 4;
            bAc.Add(MuLaw.MuLawDequantize(q, 4) * bScale);
        }

        List<double> alphaAc = [];
        List<(int, int)> alphaScanFull = [];
        int aUsable = 0;
        if (hasAlpha)
        {
            (int aNx, int aNy) = Aspect.DeriveGrid(aspect, 3);
            alphaScanFull = Dct.TriangularScanOrder(aNx, aNy);
            aUsable = Math.Min(5, alphaScanFull.Count);

            alphaAc = new List<double>(5);
            for (int i = 0; i < 5; i++)
            {
                uint q = BitPack.ReadBits(hash, bitpos, 4);
                bitpos += 4;
                alphaAc.Add(MuLaw.MuLawDequantize(q, 4) * alphaScaleVal);
            }
        }

        // Precompute scan orders
        var lScan = lScanFull.Take(lUsable).ToList();
        var lAcUsed = lAc.Take(lUsable).ToList();
        var chromaScan = chromaScanFull.Take(cUsable).ToList();
        var aAcUsed = aAc.Take(cUsable).ToList();
        var bAcUsed = bAc.Take(cUsable).ToList();
        var alphaScan = hasAlpha ? alphaScanFull.Take(aUsable).ToList() : new List<(int, int)>();
        var alphaAcUsed = alphaAc.Take(aUsable).ToList();

        // 6. Render output image
        int iw = (int)w;
        int ih = (int)h;
        byte[] rgba = new byte[iw * ih * 4];

        for (int y = 0; y < ih; y++)
        {
            for (int x = 0; x < iw; x++)
            {
                double l = Dct.DctDecodePixel(lDc, lAcUsed, lScan, x, y, iw, ih);
                double a = Dct.DctDecodePixel(aDc, aAcUsed, chromaScan, x, y, iw, ih);
                double b = Dct.DctDecodePixel(bDc, bAcUsed, chromaScan, x, y, iw, ih);
                double alpha = hasAlpha
                    ? Dct.DctDecodePixel(alphaDcVal, alphaAcUsed, alphaScan, x, y, iw, ih)
                    : 1.0;

                double lClamped = Clamp01(l);
                double[] gamutClamped = Color.SoftGamutClamp(lClamped, a, b);
                double[] rgbLinear = Color.OklabToLinearSrgb(gamutClamped);
                int idx = (y * iw + x) * 4;
                rgba[idx] = Color.LinearToSrgb8(Clamp01(rgbLinear[0]));
                rgba[idx + 1] = Color.LinearToSrgb8(Clamp01(rgbLinear[1]));
                rgba[idx + 2] = Color.LinearToSrgb8(Clamp01(rgbLinear[2]));
                rgba[idx + 3] = (byte)RoundHalfAwayFromZero(255.0 * Clamp01(alpha));
            }
        }

        return (w, h, rgba);
    }

    /// <summary>Extract the average color from a ChromaHash without full decode. Per spec §11.2.</summary>
    public static byte[] AverageColor(byte[] hash)
    {
        ulong header = 0;
        for (int i = 0; i < 6; i++)
            header |= ((ulong)hash[i]) << (i * 8);

        uint lDcQ = (uint)(header & 0x7F);
        uint aDcQ = (uint)((header >> 7) & 0x7F);
        uint bDcQ = (uint)((header >> 14) & 0x7F);
        bool hasAlpha = ((header >> 46) & 1) == 1;

        double lDc = lDcQ / 127.0;
        double aDc = (aDcQ - 64.0) / 63.0 * MaxChromaA;
        double bDc = (bDcQ - 64.0) / 63.0 * MaxChromaB;

        double lClamped = Clamp01(lDc);
        double[] gamutClamped = Color.SoftGamutClamp(lClamped, aDc, bDc);
        double[] rgbLinear = Color.OklabToLinearSrgb(gamutClamped);

        double alpha = hasAlpha ? BitPack.ReadBits(hash, 48, 5) / 31.0 : 1.0;

        return
        [
            Color.LinearToSrgb8(Clamp01(rgbLinear[0])),
            Color.LinearToSrgb8(Clamp01(rgbLinear[1])),
            Color.LinearToSrgb8(Clamp01(rgbLinear[2])),
            (byte)RoundHalfAwayFromZero(255.0 * Clamp01(alpha)),
        ];
    }
}
